keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RequiredActionLinkProvider.java

```diff
@@ -0,0 +1,212 @@
+package org.sunbird.keycloak.rest;
+
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Produces;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.Status;
+import javax.ws.rs.core.UriBuilder;
+import org.apache.commons.lang3.StringUtils;
+import org.jboss.logging.Logger;
+import org.keycloak.authentication.actiontoken.execactions.ExecuteActionsActionToken;
+import org.keycloak.common.util.Time;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.utils.KeycloakModelUtils;
+import org.keycloak.protocol.oidc.utils.RedirectUtils;
+import org.keycloak.services.ErrorResponse;
+import org.keycloak.services.managers.AppAuthManager;
+import org.keycloak.services.managers.AuthenticationManager.AuthResult;
+import org.keycloak.services.resource.RealmResourceProvider;
+import org.keycloak.services.resources.LoginActionsService;
+import org.sunbird.keycloak.utils.Constants;
+
+public class RequiredActionLinkProvider implements RealmResourceProvider {
+
+  private static Logger logger = Logger.getLogger(RequiredActionLinkProvider.class);
+  private KeycloakSession session;
+
+  public RequiredActionLinkProvider(KeycloakSession session) {
+    this.session = session;
+  }
+
+  /**
+   * Generate user required action link. The supported required actions links are for update password and verify email.
+   *
+   * @param request Request to generate required action link. The request contains following attributes:
+   *                    redirectUri: Redirect URI after performing required action
+   *                    clientId: Client ID
+   *                    requiredAction: Either UPDATE_PASSWORD or VERIFY_EMAIL
+   *                    userName: User name
+   * 
+   * @return Response containing generated required action link or error in case of failure.
+   */
+  @POST
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response generateRequiredActionLink(Map<String, String> request) {
+    logger.debug("RestResourceProvider:generateRequiredActionLink: called ");
+
+    checkRealmAdminAccess();
+    
+    String redirectUri = request.get(Constants.REDIRECT_URI);
+    String clientId = request.get(Constants.CLIENT_ID);
+    String actionName = request.get(Constants.REQUIRED_ACTION);
+    String userName = request.get(Constants.USERNAME);
+    
+    validateClientId(clientId);
+   
+    UserModel user = getUserByUserName(userName);
+    
+    ClientModel client = session.getContext().getRealm().getClientByClientId(clientId);
```

Comment:
the pattern being used for fetching the client vs fetching the user seem to be different.

For the client the code first calls a method to validate the Id, then fetches the object.
For the user, the code calls a method which validates the id and fetches the object in one call. 

Suggest choosing one of the two approaches and applying it consistently. Possibly update the `validateClientId` method  by renaming it to `getClientByClientIdOrError()`